### ISSUE TITLE
Flickering rendering on radeon

### ISSUE THREAD

**Issue by RAOF** [canonical/mir#2977]
In miriway, it seems the windows-behind-the-frontmost-flicker-through bug is back, very reproducibly and obviously.

**Comment by RAOF** [comment]
Hm. Applies to both main and 2b612b7349a114b8664f7377bb6330711d5ee035 (which I *thought* I had previously used, successfully)

**Comment by RAOF** [comment]
This turns out to be [a Mesa bug](https://gitlab.freedesktop.org/mesa/mesa/-/issues/9885), where the AMD driver returns old buffers from `gbm_surface_lock_front_buffer`; setting `mesa_glthread=false` in the environment results in the driver returning correct buffers.

**Comment by Saviq** [comment]
What can/should we do about this? Why are other compositors not affected?

**Comment by RAOF** [comment]
So, it's fixed in Mesa 24.0.2, so this is fixed in Noble (which is why it's been a while since I've seen it :wink:).

Other compositors probably don't notice this because Mir is fairly unique in *being* multithreaded. Both Mutter and Weston are eventloop-driven single threaded systems.

I'm not sure if we could set the environment variable ourselves; I don't know when the driver checks that.

**Comment by AlanGriffiths** [comment]
> I'm not sure if we could set the environment variable ourselves; I don't know when the driver checks that.

We _could_ put it in the config via `env-hacks=...`, but knowing when to do that is the thing.

**Comment by RAOF** [comment]
> > I'm not sure if we could set the environment variable ourselves; I don't know when the driver checks that.
> 
> We _could_ put it in the config via `env-hacks=...`, but knowing when to do that is the thing.

I'm not sure if that would be soon enough to ensure the driver goes down the codepath we want. If it is, we could instead do that as a quirk in the gbm-kms probe function, which would *almost* give us the right information (we'd be lacking the knowledge of whether or not the driver is new enough not to suck, and doing enough set up to *query* the driver version is almost certainly too late :disappointed: )

**Comment by RAOF** [comment]
Sadly, the driconf infrastructure does *not* allow us to specify a config option for specific *driver* versions, only application versions. I wonder if that would be a welcome patch?

**Comment by RAOF** [comment]
(We could fix this for *our* binaries by adding driconf snippets disabling the feature for our binary names)

**Comment by AlanGriffiths** [comment]
> This turns out to be [a Mesa bug](https://gitlab.freedesktop.org/mesa/mesa/-/issues/9885), where the AMD driver returns old buffers from `gbm_surface_lock_front_buffer`; setting `mesa_glthread=false` in the environment results in the driver returning correct buffers.

Would grabbing a more recent userspace (e.g. [snapping with ppa:kisak/kisak-mesa](https://github.com/AlanGriffiths/kisak-core20/blob/388bd298f8f0172c130d42efedbb713c28bef199/snap/snapcraft.yaml#L12)) be enough for Miracle & Miriway?

**Comment by RAOF** [comment]
Yes; kiask-mesa has 24.0.5, which contains the bug fix

**Comment by Saviq** [comment]
> Sadly, the driconf infrastructure does not allow us to specify a config option for specific driver versions, only application versions. I wonder if that would be a welcome patch?

Yeah having a quirk per driver version and application sounds like a good approach. But won't really help, since that'd only be available for DRI that already has the bug fixed? ;)

I think we should only do _something_ for our snaps, and that sounds like a `mesa-core22` `hwe` (hardware enablement) track, to which I alluded to in the [GNOME and KDE migration](https://forum.snapcraft.io/t/call-for-testing-ubuntu-frame-mir-2-16-4-bugfix-update-mesa-core22-minor-version-update/39715). But quirking it in `latest` would also be interesting.

I don't think we can/should go to kisak for the `latest/stable` channel - but can direct affected people to a different track.

**Comment by AlanGriffiths** [comment]
> I think we should only do _something_ for our snaps, and that sounds like a `mesa-core22` `hwe` (hardware enablement) track, to which I alluded to in the [GNOME and KDE migration](https://forum.snapcraft.io/t/call-for-testing-ubuntu-frame-mir-2-16-4-bugfix-update-mesa-core22-minor-version-update/39715). But quirking it in `latest` would also be interesting.

Do we need to "do _something_" for our `core22` snaps? In a month or two we will be moving our focus to `core24` and the corresponding archive already has this fixed. I've already got a branch of Miriway building for `core24` and can point folks to that. And, outside of Miriway and Miracle does anyone care?

> I don't think we can/should go to kisak for the `latest/stable` channel - but can direct affected people to a different track.

I agree, `latest` should track the LTS archive. If a similar situation arises in future, then a "kisak channel" is a potential solution. But I'm not sure we need it. (Yet.)

**Comment by Saviq** [comment]
> Do we need to "do _something_" for our `core22` snaps?

I suppose it's enough to have a plan. We are in the good place that we have 24 coming real soon.
